### PR TITLE
ESAPI port

### PR DESCRIPTION
**Pull request by joshuagl**
Here's my mostly complete initial port to ESAPI.

Before we can land this in master we need to resolve the issue Bill described on the mailing list with regards transient handles: https://lists.01.org/pipermail/tpm2/2018-November/000971.html

Once  this has landed in master there are various low-hanging fruit we can work on to improve the codebase following the port:
* we should be able to remove much of the crypo handling code, because ESAPI handles a lot of this for us
* we should be able to drastically reduce the session handling logic, we can use ESYS_TR_Serialize() to serialise an ESYS_TR to disk
* Probably more

### CODE REVIEW

**Comment by joshuagl** [comment]
Whoops, scan-build found 8 bugs that need addressing too...

**Comment by joshuagl** [comment]
Rebased on lastest master and resolved scan-build reported issues. Unfortunately several integration tests are failing under clang, so I'll need to reproduce locally and debug next week.

**Comment by joshuagl** [comment]
The clang branch is seeing the same issue as reported in #390, which I think is an issue in the older OpenSSL in our container.

File: tools/tpm2_startauthsession.c [hunk 1]
```diff
@@ -120,21 +120,13 @@ int tpm2_tool_onrun(ESYS_CONTEXT *ectx, tpm2_option_flags flags) {
...
     }
 
-    ESYS_TR session_handle = tpm2_session_get_handle(s);
-    TPMI_SH_AUTH_SESSION tpm_handle;
-    bool result = tpm2_util_esys_handle_to_sys_handle(ectx, session_handle,
-                    &tpm_handle);
-    if (!result) {
-        goto out;
+    if (!ctx.output.path || ctx.output.path[0] == '\0') {
+        ctx.output.path = "session.dat";
```
Comment (williamcroberts) [inline tools/tpm2_startauthsession.c]: session.ctx

File: man/tpm2_load.1.md [hunk 2]
```diff
@@ -35,8 +39,9 @@ into the TPM.
   * **-n**, **--name**=_NAME\_DATA\_FILE_:
     An optional file to save the name structure of the object.
 
-  * **-o**, **--out-context**=_CONTEXT\_FILE_:
-    An optional file to save the object context to.
+  * **-o**, **--out-context**=_CONTEXT\_FILE\_NAME_:
+    The file name of the saved object context, optional. If unspecified defaults
+    to *load.ctx*.
```
Comment (williamcroberts) [inline man/tpm2_load.1.md]: object.ctx

File: man/tpm2_loadexternal.1.md [hunk 3]
```diff
@@ -19,11 +19,13 @@ TPM. The command allows loading of a public area or both a public and a
 sensitive area.
 
 # Output
-The tool outputs the handle and name of the loaded object in a YAML format.
+The tool outputs the name of the loaded object in a YAML format and saves a
+context file for future interactions with the object. The context file name
+defaults to *external.ctx* and can be specified with **-o**.
```
Comment (williamcroberts) [inline man/tpm2_loadexternal.1.md]: object.ctx

The problem will be clobbering something that's there.... this even occurs in the current one if they don't specify the name. Perhaps a routine that looks at what's there and does something like:
```
object.ctx
object1.ctx
object2.ctx
```

Comment (joshuagl) [reply man/tpm2_loadexternal.1.md]: Great idea, I've got this mostly coded up but will do some more testing in the new year before I push here.

Comment (joshuagl) [reply man/tpm2_loadexternal.1.md]: Implemented in fc6484a

**Comment by joshuagl** [comment]
Ugh. Two strict-overflow failures on GCC.

Fortunately the tests all pass on clang with a switch from openssl to gnucrypt for the TSS crypto backend, more work to be done next week but this is very close.

**Comment by williamcroberts** [comment]
Cool. I wonder if now that the build matrix is clang + gnucrypt, if we can
Re-enable the test for getmanufec.

**Comment by joshuagl** [comment]
Sadly not, the leak in getmanufec is in code where we call OpenSSL directly, so switching the TSS to gnucrypt won't avoid it.

**Comment by joshuagl** [comment]
getmanufec leak has been addressed elsewhere in #1270 

The wider leaks causing clang issues may be in the TSS, I've filed tpm2-software/tpm2-tss#1239 to track this and implemented a change to use the gcrypt backend when building the TSS with CC=clang (6b591af).

At this point I'd really like to get this branch merged into master and iterate once there.

**Comment by joshuagl** [comment]
I've found the source of the leak and proposed a fix in tpm2-software/tpm2-tss#1242. This branch has been rebased to remove the switch to gcrypt and add some minor code quality enhancements.

CI will fail but I think it's good to go once the proposed tpm2-tss fix lands ... 🤞

**Pull request merged**